account_invoice_margin/models/account_invoice.py

```diff
@@ -5,6 +5,54 @@
 from odoo.addons import decimal_precision as dp
 
 
+class AccountInvoice(models.Model):
+    _inherit = 'account.invoice'
+
+    margin = fields.Monetary(
+        string='Margin',
+        compute='_compute_margin',
+        digits=dp.get_precision('Product Price'),
+        store=True,
+        currency_field='currency_id',
+    )
+
+    margin_signed = fields.Monetary(
+        string='Margin Signed',
+        compute='_compute_margin',
+        digits=dp.get_precision('Product Price'),
+        store=True,
+        currency_field='currency_id',
+    )
+
+    margin_percent = fields.Float(
+        string='Margin (%)',
+        digits=dp.get_precision('Product Price'),
+        compute='_compute_margin',
+        store=True,
+    )
+
+    @api.multi
+    @api.depends(
+        'invoice_line_ids.margin',
+        'invoice_line_ids.margin_signed',
+        'invoice_line_ids.price_subtotal',
+    )
+    def _compute_margin(self):
+        for invoice in self:
```

Comment:
Please do this through `read_group` for performance.

Reply:
Hi @pedrobaeza. thanks for your review.

sorry, I'm not sure to understand. Do you want me to write something like : 

    AccountInvoiceLine.read_group(
        [('invoice_id', 'in', self.ids)],
        ['margin', 'margin_signed', 'price_subtotal'],
        'invoice_id'
    )

Reply:
Yes, that's it, and you get directly the sum. Take care that read_group syntax is not exactly that, but you got the point.

Reply:
Hi I tried this code but, the onchange is so failing. (when I change a line, the invoice margin is set to 0)
Do you know how to fix it ? Or did I missed something ?

        for invoice in self:
            res = self.env['account.invoice.line'].read_group(
                [('invoice_id', '=', self.id)],
                ['margin', 'margin_signed', 'price_subtotal'],
                ['invoice_id'],
            )
            if res:
                res = res[0]
                invoice.update({
                    'margin': res['margin'],
                    'margin_signed': res['margin_signed'],
                    'margin_percent': res['price_subtotal'] and (
                        res['margin_signed'] / res['price_subtotal'] * 100),
                })

The problem I see with what you propose is that self.ids (or self.id) doesn't have allways a value.

Reply:
You have to use `res.get('field', 0)`

Reply:
That's not the problem. The problem is : 

- Open a draft invoice
- change a value in a line
- the function is called but  ``res == []``
-> the values are set to 0 -> KO

This behaviour does'nt occure with the current code of the PR.

Reply:
You can add a special case for NewId, but the performance gets really improved with this for the 99% percentage of cases (stored data).

Reply:
OK ! It works ! 

Better ? : https://github.com/OCA/margin-analysis/pull/68/files#diff-53256610b917670c26058816b8fcaac1R41

Reply:
Not really:

- you are still losing the performance doing N `read_group` calls: do only once with all the real IDs together.
- You are not computing for the virtual IDs via the normal computation (`sum('invoice_line_ids.margin...)`).

Reply:
Do you want me to do a PR over your branch with this?

Reply:
Yes, please. Sorry for not understanding.
I made some grep in other repo, but did not found read_group in other compute method, to have inspiration.